### PR TITLE
[Feature] Implement server IP bans

### PR DESCRIPTION
**Pull request by zquestz**
Implements IP bans on the server. The criteria are:

- The user gets banned from a pool (requires all other pool participants to agree), and the ban score goes up by 1. (We decrement this after 30 minutes)
- When the tracker sees a ban score of 3 it doesn't allow them to connect.
- Added `banned` and `banScore` keys to stats endpoint.
- Added `version` to the announce message, this should help with protocol updates that type doesn't make sense for.
- Added `version` key in each pools stats output.

This provides a safety cushion for users if they need to update their client. But also makes it a lot more robust against people that would be malicious.

Let me know what you think. =)

### CODE REVIEW

File: message/message.pb.go [hunk 1]
```diff
@@ -355,8 +355,9 @@ func (m *Address) GetAddress() string {
 }
 
 type Registration struct {
-	Amount uint64      `protobuf:"varint,1,opt,name=amount" json:"amount,omitempty"`
-	Type   ShuffleType `protobuf:"varint,2,opt,name=type,enum=ShuffleType" json:"type,omitempty"`
+	Amount  uint64      `protobuf:"varint,1,opt,name=amount" json:"amount,omitempty"`
+	Type    ShuffleType `protobuf:"varint,2,opt,name=type,enum=ShuffleType" json:"type,omitempty"`
+	Version uint64      `protobuf:"varint,3,opt,name=version" json:"version,omitempty"`
```
Comment (emergent-reasons) [inline message/message.pb.go]: Sorry if it wasn't cashshuffle, but I remember somebody talking about missing message fields causing a problem. Is `omitempty` intentional / not a problem here?

**Comment by emergent-reasons** [comment]
In register.go, do we not need `version: registration.Version,` in the `trackerData`?

File: server/tracker.go [hunk 2]
```diff
@@ -3,25 +3,45 @@ package server
...
+const (
+	// banTime is the amount of time to ban an IP.
+	banTime = 30 * time.Minute
+
+	// banScoreTick is the ban score increment on each pool ban.
+	banScoreTick = 1
+
+	// maxBanScore is the score the connection much reach to
+	// be banned by IP.
+	maxBanScore = 3
```
Comment (cculianu) [inline server/tracker.go]: 3 strikes and you're out, eh? Ha!

**Comment by cculianu** [comment]
Interesting.  So there was already ban logic in there -- it just didn't actually ban the player from connecting (perhaps just banned them from that particular pool?).

Cool.  This looks good on quick inspection.  Looks like it should work as intended.

I'm not a go programmer anymore (it's been literally years) -- so don't take my approval as a "thorough review" -- more a simple "yeah, the logic looks like what I was expecting" kind of thing.  👍

File: message/message.pb.go (hunk 1 above)
Comment (zquestz) [reply message/message.pb.go]: This is fine for this use case. The default value is always generated client side if omitted in the protobuf stream.

File: server/tracker.go (hunk 2 above)
Comment (zquestz) [reply server/tracker.go]: Yep =)

**Comment by zquestz** [comment]
@emergent-reasons great catch on the registration message. This has been fixed. Coding at 2AM has its side effects. =)

File: server/tracker.go [hunk 3]
```diff
@@ -94,6 +117,56 @@ func (t *Tracker) count() int {
...
 
+// bannedIP returns true if the player has been banned from the server.
+func (t *Tracker) bannedIP(conn net.Conn) bool {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	ip, _, _ := net.SplitHostPort(conn.RemoteAddr().String())
+
+	banData := t.bannedIPs[ip]
+	if banData != nil && banData.score >= maxBanScore {
```
Comment (eyeamera) [inline server/tracker.go]:
```suggestion
	if banData, ok := t.bannedIPs[ip]; ok && banData.score >= maxBanScore {
```

File: server/tracker.go [hunk 4]
```diff
@@ -94,6 +117,56 @@ func (t *Tracker) count() int {
...
+	time.Sleep(banTime)
+
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if _, ok := t.bannedIPs[ip]; ok {
+		t.bannedIPs[ip].score -= banScoreTick
+	}
+
+	if t.bannedIPs[ip].score == 0 {
```
Comment (eyeamera) [inline server/tracker.go]: This might not exist? Maybe include this in the `ok` check block from above?

File: server/tracker.go (hunk 3 above)
Comment (zquestz) [reply server/tracker.go]: Fixed.

File: server/tracker.go (hunk 4 above)
Comment (zquestz) [reply server/tracker.go]: Fixed.

File: message/message.pb.go (hunk 1 above)
Comment (emergent-reasons) [reply message/message.pb.go]: Thanks for explaining. I should have known that :) Whatever I remember must have been somewhere else.

**Pull request merged**